logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -917,4 +927,39 @@ private Batch deserialize() {
             return new Batch(elements, firstSeqNum, Queue.this);
         }
     }
+
+    private boolean awaitReadDemand(final long timeoutMillis, final int elementsNeeded) throws InterruptedException {
+        assert this.lock.isHeldByCurrentThread();
+
+        final long deadlineMillis = Math.addExact(System.currentTimeMillis(), timeoutMillis);
+        this.readDemand = new ReadDemand(deadlineMillis, elementsNeeded);
+
+        boolean unElapsed = this.notEmpty.awaitUntil(new Date(deadlineMillis));
+        this.readDemand = null;
```

Comment:
I don't have big concern but wanna clarify the behaviour.

Reader releases the lock when `notEmpty.awaitUntil()` is called and do the waiting. Potentially, other readers could acquire the lock and update `this.readDemand` to a further delay `deadlineMillis`. Updating `this.readDemand` is not atomic and I believe it is not required to be atomic. The worst case is the deadline extends to (timeout 50ms x number of threads).

Is my understanding correct? If so, can we add a comment to remind that the action is not atomic?

Reply:
Other readers cannot acquire the lock because `Queue#readSerializedBatch` (the only place where a reader can acquire the queue's lock) is `synchronized`.